docs/DeveloperGuide.md

```diff
@@ -133,89 +133,188 @@ Classes used by multiple components are in the `seedu.addressbook.commons` packa
+
+#### Implementation
 
-### \[Proposed\] Data archiving
+The command recall mechanism is designed around maintaining a list of previously executed commands, and traversing this list
+when the user requests to recall a previous command.
 
-_{Explain here how the data archiving feature will be implemented}_
+The command recall mechanism is facilitated by `CommandHistory`, which consists of a list of previously executed commands,
+stored internally as a list of `String`, and a `pointer` to determine the location of the currently recalled command.
+In addition, it also consists of a cache, which is used to store the user's existing input.
+
+<div markdown="span" class="alert alert-info">
+:information_source: **Note:** The data stored in `CommandHistory` does not persist when the application is exited.
+</div>
+
+When the application is started, the `CommandHistory` is initialized with an empty list, and the `pointer` is set to
+the size of the list, currently zero. Every successful command executed will be appended to the list and the `pointer` will be set
+to the current size of the list after the update, regardless of its original position.
+
+When the user presses the <kbd>UP</kbd> arrow key, a check is performed to determine if there is a next available command to recall.
+If there is a previous command available, `CommandHistory` next checks if the `pointer` is currently pointing to an entry in its list.
+If the `pointer` is not currently pointing to an entry in the list, i.e. its index is one greater than the last entry in the list,
+the current text in the command box is stored in the cache. Otherwise, there is no change to the cached value.
+Subsequently, `CommandHistory` decreases its `pointer` by one, and the respective `String` in the list is returned.
+
+The text in the command box is then set to the returned `string`. As such, the `pointer` always points to the command that is displayed
+in the command box, and if the text in the command box is a new command, then the `pointer` index is one greater than the last stored command.
+
+If there are no earlier commands to recall, then there is no change to the text in the command box.
+
+<div markdown="span" class="alert alert-info">
+:information_source: **Note:** Any edits made by the user to recalled commands are not stored.
+</div>
+
+When the user presses the <kbd>DOWN</kbd> arrow key, a check is performed to determine if there is a next available command to recall.
+If there is a next available command, the `pointer` index is increased and the respective `String` in the list is returned.
+
+![CommandHistoryNoNext](images/CommandHistoryReturnsCache.png)
+
+Otherwise, if there is no available next command, i.e. the pointer index is at the last element in the list, then the cached value (if there is one) is returned instead.
+
+![CommandHistoryActivityDiagram](images/CommandHistoryActivityDiagram.png)
+
+The activity diagram above provides a summary of the recall command mechanism.
```

Comment:
Would it be better to shift this diagram to around line 278? The explanation is rather dense there.

Reply:
I'll be inserting another diagram over there (around line 280) to explain the case when the current value is cached in a subsequent PR. So I think I'll leave the summary at the end for now.

Reply:
Sure!